spring-integration-core/src/main/java/org/springframework/integration/expression/ExpressionEvalMap.java

```diff
@@ -206,22 +207,35 @@ public static ExpressionEvalMapBuilder from(Map<String, ?> expressions) {
 	 */
 	public static class ComponentsEvaluationCallback implements EvaluationCallback {
 
+		@Nullable
 		private final EvaluationContext context;
 
+		@Nullable
 		private final Object root;
 
+		private final boolean rootExplicitlySet;
+
+		@Nullable
 		private final Class<?> returnType;
 
-		public ComponentsEvaluationCallback(EvaluationContext context, Object root, Class<?> returnType) {
+		public ComponentsEvaluationCallback(@Nullable EvaluationContext context, @Nullable Object root,
+				boolean rootExplicitlySet, @Nullable Class<?> returnType) {
+
 			this.context = context;
 			this.root = root;
+			this.rootExplicitlySet = rootExplicitlySet;
```

Comment:
It's not clear what the purpose of this is; we still get an NPE if the root has explicitly been set to `null`.

>is that we need to deal with provided rootObject even if it is `null`

When I changed the HRHES to `.withRoot(null)` BOOM!!

Reply:
That's correct: it is exactly what SpEL contract does in the `getValue(EvaluationContext context, @Nullable Object rootObject)`.
If you explicitly provide a root as `null`, then it is used as is, without any consultation with `context`.

So, we fix an NPE on our level because we really have a root, but we still don't break SpEL contract when root is explicitly provided as `null`

Reply:
So you are saying the NPE is ok,  because it's in "user" code (in this case, the lambda) so the user is responsible for a null check?

Reply:
That's right if (s)he provides a `null` explicitly, but in our case we just miss the proper root to propagate.
See all the `Expression.getValue()` with `root` - it is `@Nullable`. Therefore without checking for `null` it is really going to lead into NPE.
But in our case we provide all the stuff for expression environment, therefore it is safe not check because we guarantee now a non-null root.
See changes in those HTTP components.